### PR TITLE
Created documentation generation using Sphinx and ReadTheDocs

### PR DESCRIPTION
**Pull request by dolphingarlic**
This is for that one GCI task

### CODE REVIEW

File: apertium/__init__.py [hunk 1]
```diff
@@ -12,7 +12,8 @@ class ModeNotInstalled(ValueError):
     pass
 
 
-def update_modes(pair_path):  # type: (str) -> None
+def update_modes(pair_path):
+    """type: (str) -> None"""
```
Comment (sushain97) [inline apertium/__init__.py]: Does mypy process type annotations in comments like this? Please verify that it does and show me the error message.

Comment (dolphingarlic) [reply apertium/__init__.py]: It works there because the docstrings are different, I think. Sphinx doesn't detect comments as docstrings.

Comment (dolphingarlic) [reply apertium/__init__.py]: That's why something like the streamparser.mainpos method doesn't show the parameter or return types

Comment (dolphingarlic) [reply apertium/__init__.py]: Should I just convert everything to mypy docstrings?

Comment (sushain97) [reply apertium/__init__.py]: Maybe something like https://github.com/agronholm/sphinx-autodoc-typehints exists for the comment version. This diff ruins all the typechecking though so it definitely isn't what we want.

Comment (sushain97) [reply apertium/__init__.py]: There's no such thing as a mypy docstring afaik. Mypy has type annotations via the comment syntax or the newer syntax.

Comment (dolphingarlic) [reply apertium/__init__.py]: Using something like this: https://pypi.org/project/sphinx-autodoc-typehints/ ?

Comment (dolphingarlic) [reply apertium/__init__.py]: Also, should I be documenting those private methods or should I just leave those out?

**Comment by dolphingarlic** [comment]
Should I be worrying about these Travis CI issues?

File: apertium/__init__.py (hunk 1 above)
Comment (sushain97) [reply apertium/__init__.py]: There's no need to document private interfaces.

**Comment by sushain97** [comment]
I don't merge PRs that fail the build if it was passing before unless it's not your fault :)

**Comment by dolphingarlic** [comment]
Here's what I get when I run "mypy -p apertium --ignore-missing-imports --txt-report coverage" on the files with the docstrings:


Mypy Type Check Coverage Summary
================================

Script: index

+----------------------+-------------------+---------+
| Module               | Imprecision       | Lines   |
+----------------------+-------------------+---------+
| apertium             |   0.00% imprecise |  46 LOC |
| apertium.analysis    |   8.86% imprecise |  79 LOC |
| apertium.generation  |   0.00% imprecise |  65 LOC |
| apertium.mode_search |   0.00% imprecise |  88 LOC |
| apertium.translation |   5.85% imprecise | 188 LOC |
| apertium.utils       |   1.49% imprecise |  67 LOC |
+----------------------+-------------------+---------+
| Total                |   3.56% imprecise | 533 LOC |
+----------------------+-------------------+---------+

File: apertium/__init__.py [hunk 2]
```diff
@@ -25,7 +30,12 @@ def update_modes(pair_path):  # type: (str) -> None
             generators[lang_pair] = (dirpath, modename)
 
 
-def append_pair_path(pair_path):  # type: (str) -> None
+def append_pair_path(pair_path):
```
Comment (sushain97) [inline apertium/__init__.py]: This is a private method so we can remove the docstrings here.

File: apertium/__init__.py [hunk 3]
```diff
@@ -12,7 +12,12 @@ class ModeNotInstalled(ValueError):
     pass
 
 
-def update_modes(pair_path):  # type: (str) -> None
+def update_modes(pair_path):
+    # type: (str) -> None
```
Comment (sushain97) [inline apertium/__init__.py]: I prefer the type on the same line as it was previously. Is that not possible?

File: apertium/__init__.py (hunk 2 above)
Comment (sushain97) [reply apertium/__init__.py]: (it should be prefixed with an underscore)

File: apertium/analysis/__init__.py [hunk 4]
```diff
@@ -1,43 +1,84 @@
-from streamparser import parse, LexicalUnit  # noqa: F401
+from streamparser import parse, LexicalUnit  # type: ignore  # noqa: F401
 
-import apertium
-from apertium.utils import to_alpha3_code, execute, parse_mode_file
+import apertium  # type: ignore
```
Comment (sushain97) [inline apertium/analysis/__init__.py]: Why are we sprouting this `type: ignore`s everywhere? It seemed to work before without any of them...? By doing this, you're effectively nulling half of what mypy does.

File: apertium/utils.py [hunk 5]
```diff
@@ -1,25 +1,41 @@
+from apertium.iso639 import iso_639_codes  # type: ignore
```
Comment (sushain97) [inline apertium/utils.py]: The previous import sort order was intentional.

File: apertium/translation/__init__.py [hunk 6]
```diff
@@ -1,27 +1,57 @@
+from apertium.utils import to_alpha3_code, execute, parse_mode_file  # type: ignore  # noqa: F401
```
Comment (sushain97) [inline apertium/translation/__init__.py]: The previous import sort order was intentional.

File: docs/make.bat [hunk 7]
```diff
@@ -0,0 +1,35 @@
+@ECHO OFF
```
Comment (sushain97) [inline docs/make.bat]: uhh, the script committed to this repo should be a shell script, not DOS batch.

File: docs/source/index.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
+Apertium + Python
```
Comment (sushain97) [inline docs/source/index.md]: This should 100% not be copied like this. We need to either use a symlink or reference the README directly.

Comment (sushain97) [reply docs/source/index.md]: Similarly below. Maybe the README should just be split.

File: apertium/__init__.py (hunk 3 above)
Comment (dolphingarlic) [reply apertium/__init__.py]: It is possible. I'll change it back

File: docs/make.bat (hunk 7 above)
Comment (dolphingarlic) [reply docs/make.bat]: That was automatically generated by Sphinx (for Windows development or something). There already is a shell script, I think. I'll just delete the file for now.

File: docs/source/index.md (hunk 8 above)
Comment (dolphingarlic) [reply docs/source/index.md]: Should I create symlinks for all the .md files?

File: apertium/analysis/__init__.py (hunk 4 above)
Comment (dolphingarlic) [reply apertium/analysis/__init__.py]: mypy would give me errors for imports before. I'll take away those comments

Comment (sushain97) [reply apertium/analysis/__init__.py]: If mypy had errors, the build wouldn't have passed?

File: docs/make.bat (hunk 7 above)
Comment (sushain97) [reply docs/make.bat]: Sounds good. Please verify that there is a shell script.

File: docs/source/index.md (hunk 8 above)
Comment (sushain97) [reply docs/source/index.md]: Sounds good. The usage markdown file can go inside the `docs` directory.

Comment (dolphingarlic) [reply docs/source/index.md]: This is the index.md for the docs generation though

Comment (dolphingarlic) [reply docs/source/index.md]: I made the docs only about the code now

Comment (dolphingarlic) [reply docs/source/index.md]: I couldn't make the symlinks. reCommonMark doesn't recognize them for some reason

File: apertium/analysis/__init__.py (hunk 4 above)
Comment (dolphingarlic) [reply apertium/analysis/__init__.py]: Maybe it was just my computer. Anyways, I changed them back now

File: docs/source/index.md (hunk 8 above)
Comment (dolphingarlic) [reply docs/source/index.md]: Do you think I should make a separate page for each module?

File: apertium/__init__.py [hunk 9]
```diff
@@ -25,7 +29,7 @@ def update_modes(pair_path):  # type: (str) -> None
             generators[lang_pair] = (dirpath, modename)
 
 
-def append_pair_path(pair_path):  # type: (str) -> None
+def _append_pair_path(pair_path):  # type: (str) -> None
```
Comment (sushain97) [inline apertium/__init__.py]: Whoops, I misspoke earlier. `update_modes` is 'private', not `append_pair_path`. Mind switching that?

File: coverage/index.txt [hunk 10]
```diff
@@ -0,0 +1,17 @@
+Mypy Type Check Coverage Summary
```
Comment (sushain97) [inline coverage/index.txt]: Please delete this.

File: docs/index [hunk 11]
```diff
@@ -0,0 +1,41 @@
+Apertium + Python
```
Comment (sushain97) [inline docs/index]: Does this file not have a file extension?

File: docs/about [hunk 12]
```diff
@@ -0,0 +1,21 @@
+About
```
Comment (sushain97) [inline docs/about]: Should this be `about.md`?

File: docs/about [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+-----------------------------
+
+- The existing code base has the subprocess implementation of the basic functions of Apertium.
+- A branch called `windows` has the implementation for the `windows` support and will soon be available on master. Detailed instructions can be found [here](https://gist.github.com/arghyatiger/c8aab476022158f4bdb3dbe45308cdb4)
+
+Contribute
+----------
+
+- Issue Tracker: <https://www.github.com/apertium/apertium-python/issues>
+- Source Code: <https://www.github.com/apertium/apertium-python>
```
Comment (sushain97) [inline docs/about]: Please always use newlines at the end of a file.

File: docs/source/about.md [hunk 14]
```diff
@@ -0,0 +1,21 @@
+About
```
Comment (sushain97) [inline docs/source/about.md]: Actually, why are there two of these?

File: docs/usage [hunk 15]
```diff
@@ -0,0 +1,62 @@
+Usage
```
Comment (sushain97) [inline docs/usage]: Why is this duplicated?

Comment (dolphingarlic) [reply docs/usage]: I tried to create a symlink, but I deleted it now

File: docs/source/about.md (hunk 14 above)
Comment (dolphingarlic) [reply docs/source/about.md]: I tried to create a symlink, but I deleted it now

File: docs/about (hunk 12 above)
Comment (dolphingarlic) [reply docs/about]: That was the file used to create a symlink

File: apertium/__init__.py (hunk 9 above)
Comment (dolphingarlic) [reply apertium/__init__.py]: I did that now

File: coverage/index.txt (hunk 10 above)
Comment (dolphingarlic) [reply coverage/index.txt]: I deleted it now

File: docs/index (hunk 11 above)
Comment (dolphingarlic) [reply docs/index]: That was the file used to create a symlink

**Pull request merged**